File: python/p0002_add_two_numbers.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        p = head = ListNode(0)
        carry = 0

        while l1 or l2:
            # calculate sum
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            total = x + y + carry
            total, carry = total % 10, total // 10
            p.next = ListNode(total)

            # proceed to next node
            p = p.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        if carry:
            p.next = ListNode(1)

        return head.next or head

File: src/solutions/p0002_add_two_numbers.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut head: &mut ListNode = &mut dummy;
        let mut l1 = l1.as_ref();
        let mut l2 = l2.as_ref();
        let mut carry = 0;

        while l1.is_some() || l2.is_some() {
            let mut sum = carry;
            if let Some(n1) = l1 {
                sum += n1.val;
                l1 = n1.next.as_ref();
            }
            if let Some(n2) = l2 {
                sum += n2.val;
                l2 = n2.next.as_ref();
            }
            carry = sum / 10;
            head.next = Some(Box::new(ListNode::new(sum % 10)));
            head = head.next.as_mut().unwrap();
        }

        if carry > 0 {
            head.next = Some(Box::new(ListNode::new(1)));
        }

        dummy.next
    }
}
